### PR TITLE
Avoid endless page requesting if when failing requests + manual retry

### PR DESCRIPTION
**Pull request by antgamdia**
### Description of the change

This PR aims at solving one of the issues detected in https://github.com/kubeapps/kubeapps/issues/2484#issuecomment-788658490. Currently, the pagination is trying to increase the page regardless of the status code.

In this PR we stop requesting more pages if an error is detected. However, solely adding that has a severe impact on the UX: if request happens to fail, all the previously requested items will be discarded and the only way to retry is by performing a hard reload (f5).
Wherefore this PR also adds a mechanism to trigger: i) an increase of the requested page; 2) clearing the current error.

### Benefits

 A live example is depicted below:

tryAgainReq

### Possible drawbacks

This approach just assumes there is a problem on a given page and we force to request the next one. As a result, the failing page will remain unrequested. 

### Applicable issues

  - related https://github.com/kubeapps/kubeapps/issues/2484

### Additional information

Perhaps 'try again' is not the best message (since we do discard the failing page), see what you think.

### CODE REVIEW

**Comment by absoludity** [comment]
> ### Description of the change
> In this PR we stop requesting more pages if an error is detected. However, solely adding that has a severe impact on the UX: if request happens to fail, all the previously requested items will be discarded and the only way to retry is by performing a hard reload (f5).

Great that you avoided that. I don't have the full context but is there a reason we need or should be requesting further pages before we've successfully received the results of the previous ones? I didn't think people would scroll that fast - if we only ever requested one page at a time (serially) we wouldn't have "all the previously requested items will be discareded" right? Or maybe I've misunderstood. 

> Wherefore this PR also adds a mechanism to trigger: i) an increase of the requested page; 2) clearing the current error.

Could we just retry the request automatically some threshold of times before failing?

> ### Benefits
> 
> A live example is depicted below:
> 
> tryAgainReq

Thanks for the visual!

> ### Possible drawbacks
> 
> This approach just assumes there is a problem on a given page and we force to request the next one. As a result, the failing page will remain unrequested.

I don't see why we force the next one - why not re-request the current one that failed? (whether or not it's done automatically).


> ### Applicable issues
> 
>     * related #2484
> 
> 
> ### Additional information
> 
> Perhaps 'try again' is not the best message (since we do discard the failing page), see what you think.

Hah - yes, though I think we shouldn't discard the failing page. We know there's not going to be a reason for a particular page to fail and not others.

**Review by absoludity — APPROVED** [review]
+1 as I think this is better than what we have, but I do wonder whether we should be requesting the pages serially (looks like it's requesting further pages before a successful response for the current). See what you think.

File: dashboard/src/components/Catalog/Catalog.tsx [hunk 1]
```diff
@@ -256,8 +257,20 @@ function Catalog(props: ICatalogProps) {
 
   // Required to have the latest value of page
   const setPageWithContext = () => {
+    if (!error) {
```
Comment (andresmgot) [inline dashboard/src/components/Catalog/Catalog.tsx:260]: as far as I can tell, just checking this error is enough to fix this issue.

**Review by andresmgot — COMMENTED** [review]
Thanks @antgamdia. I think we are fixing an error that doesn't exist here though. You are doing two things:

 - If a page request fails, don't request the next one.
 - Smooth recovery path in one (of many) pages fail.
 
IMO the second one is not really needed. It's highly improbable that if a page success to be retrieved (e.g page 1) the next one can fail. Also, as far as I can tell, we are not re-requesting the same page so we would be leaving the catalog in a inconsistent way.

I would just show the error and let people refresh the whole page (since the most probably page that can fail is the first one) but if you want to keep it this path, I would adapt this code to retry the same page failing.

**Comment by antgamdia** [comment]
> we wouldn't have "all the previously requested items will be discarded" right? 

Yes, currently the redux reducer is setting `items: []` if an error occurs. This PR changes this behavior to be `items: state.items` 

> why not re-request the current one that failed? 

You're right. I make the assumption that just a specific page is failing. However, it doesn't really make sense, as 'it's highly improbable that if a page success to be retrieved (e.g page 1) the next one can fail'.
So, I have changed the implementation to perform a re-fetch of the failing page.


> I would just show the error and let people refresh the whole page (since the most probably page that can fail is the first one) but if you want to keep it this path, I would adapt this code to retry the same page failing.

Mmm, I don't have a strong opinion here. My point on going this way is: supposing a user is fetching a repo with several pages, and, because of an eventual network issue, the last request goes wrong. As this user, I'd like to have a way to recover from this error. But, also, I agree with you, this is very unlikely to happen.

So, to sum up. I would leave this fallback mechanism implemented since users are still able to just refresh the page.

**Review by andresmgot — APPROVED** [review]
